### PR TITLE
Feature/tao 9128/select releasing branch

### PR DESCRIPTION
**Pull request by ricproenca-tao**
**Related to task:** 
https://oat-sa.atlassian.net/browse/TAO-9128

**Description:**
Create a select releasing branch step that, if `is version-to-release` is provided select the proper branch for this version, if not select the branch with biggest version.

**Notes:**
- added a new dependency 'semver-compare';

**How to test:**
- Add the proper steps to `createRelease.js` file, then;
`node .\index.js createRelease -d` or `node .\index.js createRelease -d --version-to-release XX.X.X`

**Unit tests:**
`node .\tests\unit\release\selectReleasingBranch\test.js`

### CODE REVIEW

File: src/git.js [hunk 1]
```diff
@@ -98,6 +98,23 @@ module.exports = function gitFactory(repository = '', origin = 'origin') {
...
+         * @returns {Promise}
+         */
+        checkout(branchName){
+            return git(repository).checkout(branchName);
+        },
+
+        /**
+         * Gets any new commits, references (like tags), branches and files from a remote repository
+         * @param {*} options
+         */
```
[the remark this answers is not in the document]
Comment (ricproenca-tao) [reply src/git.js]: done!

File: src/release.js [hunk 2]
```diff
@@ -438,5 +440,109 @@ module.exports = function taoExtensionReleaseFactory(baseBranch, branchPrefix, o
...
+         * - or find the biggest version and find branch with version on it
+         */
+        async selectReleasingBranch() {
+            let version;
+            let branchToRelease;
+
+            // Filter all branches to the ones that have release in the name
+            await gitClient.fetch({'--prune': true});
+            const allBranches = await gitClient.getLocalBranches();
+            const possibleBranches = allBranches.filter(branch => (branch.includes('remotes') && branch.includes('origin') && branch.includes('release')));
```
[the remark this answers is not in the document]
Comment (ricproenca-tao) [reply src/release.js]: yes. true.

File: src/release.js [hunk 3]
```diff
@@ -438,5 +440,109 @@ module.exports = function taoExtensionReleaseFactory(baseBranch, branchPrefix, o
...
             log.done(`${data.extension.name} is clean`);
         },
+
+        /**
+         * Select releasing branch
+         * - picking version-to-release CLI option and find branch with version on it
+         * - or find the biggest version and find branch with version on it
+         */
+        async selectReleasingBranch() {
+            let version;
```
[the remark this answers is not in the document]
Comment (ricproenca-tao) [reply src/release.js]: yes, but I don't want to pollute an external param.
I will leave

File: src/release.js [hunk 4]
```diff
@@ -438,5 +440,109 @@ module.exports = function taoExtensionReleaseFactory(baseBranch, branchPrefix, o
...
+            return { branch, version };
+        },
+
+        /**
+         * Get the branch with the provided semver version in the name
+         * @param {Array.<String>} possibleBranches - array of branches
+         * @param {String} versionToFind - semVer version X.X.X.X
+         * @returns {*}
+         */
+        getBranchWithVersion(possibleBranches = [], versionToFind) {
```
[the remark this answers is not in the document]
Comment (ricproenca-tao) [reply src/release.js]: You're right. Simpliest logic.

File: src/release.js [hunk 5]
```diff
@@ -470,5 +471,66 @@ module.exports = function taoExtensionReleaseFactory(params = {}) {
...
+                }
+            }
+
+            if (data.releasingBranch) {
+                log.done(`Branch ${data.releasingBranch} is selected.`);
+                return;
+            }
+
+            log.exit('Cannot find any branch with valid version.');
+        },
```
Comment (oatymart) [inline src/release.js]: When we exit, I would like to at least show the user the branch name we searched for.

I think this means you will need to exit in 2 places, with different messages.

first time `log.exit('Cannot find the branch remotes/${origin}/${branchPrefix}-${versionToRelease}');` after L490
second time `log.exit('Cannot find any branches matching remotes/${origin}/${branchPrefix}-');` after L498

File: src/release.js [hunk 6]
```diff
@@ -470,5 +471,66 @@ module.exports = function taoExtensionReleaseFactory(params = {}) {
...
+            const allBranches = await gitClient.getLocalBranches();
+
+            if (versionToRelease) {
+                const branchName = `remotes/${origin}/${branchPrefix}-${versionToRelease}`;
+                if (allBranches.includes(branchName)) {
+                    data.releasingBranch = branchName;
+                    data.version = versionToRelease;
+                }
+            } else {
+                const branchName = `remotes/${origin}/${branchPrefix}`;
```
Comment (oatymart) [inline src/release.js]: Technically it's a `partialBranchName` at this stage

File: src/release.js [hunk 7]
```diff
@@ -470,5 +471,66 @@ module.exports = function taoExtensionReleaseFactory(params = {}) {
...
+                const possibleBranches = allBranches.filter(branch => branch.includes(branchName));
```
Comment (oatymart) [inline src/release.js]: How about `.includes` -> `.startsWith`?

Comment (ricproenca-tao) [reply src/release.js]: yes, I agree

File: src/release.js (hunk 6 above)
Comment (ricproenca-tao) [reply src/release.js]: yes, ok.

File: src/release.js (hunk 5 above)
Comment (ricproenca-tao) [reply src/release.js]: 1st -> log.exit(`Cannot find the branch ${branchName}`);
2nd -> log.exit(`Cannot find any branches matching ${partialBranchName}`);

Comment (oatymart) [reply src/release.js]: great ;)

File: src/release.js [hunk 8]
```diff
@@ -470,5 +471,68 @@ module.exports = function taoExtensionReleaseFactory(params = {}) {
...
+                    data.releasingBranch = highestVersionBranch.branch;
+                    data.version = highestVersionBranch.version;
+                } else {
+                    log.exit(`Cannot find any branches matching '${partialBranchName}'.`);
+                }
+            }
+
+            if (data.releasingBranch) {
+                log.done(`Branch ${data.releasingBranch} is selected.`);
+                return;
```
Comment (oatymart) [inline src/release.js]: The `return;` is no longer needed, and I think the wrapping `if` can go too.

Comment (ricproenca-tao) [reply src/release.js]: forgot it.

**Pull request merged**